x/cork/keeper/keeper.go

```diff
@@ -0,0 +1,272 @@
+		if !seenAddresses.Add(address) {
+			continue
+		}
+
+		if handler(address) {
+			break
+		}
+	}
+}
+
+//////////////////
+// CommitPeriod //
+//////////////////
+
+// SetCommitPeriodStart sets the current vote period start height
+func (k Keeper) SetCommitPeriodStart(ctx sdk.Context, height int64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set([]byte{types.CommitPeriodStartKey}, sdk.Uint64ToBigEndian(uint64(height)))
+}
+
+// GetCommitPeriodStart returns the vote period start height
+func (k Keeper) GetCommitPeriodStart(ctx sdk.Context) (int64, bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get([]byte{types.CommitPeriodStartKey})
+	if len(bz) == 0 {
+		return 0, false
+	}
+
+	return int64(sdk.BigEndianToUint64(bz)), true
+}
+
+// HasCommitPeriodStart returns true if the vote period start has been set
+func (k Keeper) HasCommitPeriodStart(ctx sdk.Context) bool {
+	return ctx.KVStore(k.storeKey).Has([]byte{types.CommitPeriodStartKey})
+}
+
+////////////
+// Params //
+////////////
+
+// GetParamSet returns the vote period from the parameters
+func (k Keeper) GetParamSet(ctx sdk.Context) types.Params {
+	var p types.Params
+	k.paramSpace.GetParamSet(ctx, &p)
+	return p
+}
+
+// setParams sets the parameters in the store
+func (k Keeper) setParams(ctx sdk.Context, params types.Params) {
+	k.paramSpace.SetParamSet(ctx, &params)
+}
+
+/////////////////////////
+// Invalidation Nonces //
+/////////////////////////
+
+func (k Keeper) GetLatestInvalidationNonce(ctx sdk.Context) uint64 {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get([]byte{types.LatestInvalidationNonceKey})
+	return sdk.BigEndianToUint64(bz)
+}
+
+func (k Keeper) SetLatestInvalidationNonce(ctx sdk.Context, invalidationNonce uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set([]byte{types.LatestInvalidationNonceKey}, sdk.Uint64ToBigEndian(invalidationNonce))
+}
+
+func (k Keeper) IncrementInvalidationNonce(ctx sdk.Context) uint64 {
+	store := ctx.KVStore(k.storeKey)
+	nextNonce := k.GetLatestInvalidationNonce(ctx) + 1
+	store.Set([]byte{types.LatestInvalidationNonceKey}, sdk.Uint64ToBigEndian(nextNonce))
+	return nextNonce
+}
+
+///////////
+// Votes //
+///////////
+
+func (k Keeper) GetWinningVotes(ctx sdk.Context, threshold sdk.Dec) (winningVotes []types.Cork) {
+
+	var corks []types.Cork
+	var corkPowers []int64
+
+	totalPower := k.stakingKeeper.GetLastTotalPower(ctx)
+
+	k.IterateCorks(ctx, func(val sdk.ValAddress, addr common.Address, cork types.Cork) (stop bool) {
+		validator := k.stakingKeeper.Validator(ctx, val)
+		validatorPower := validator.GetConsensusPower(k.stakingKeeper.PowerReduction(ctx))
+
+		found := false
+		for i, rv := range corks {
+			if rv.Equals(cork) {
+				corkPowers[i] += validatorPower
+
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			corks = append(corks, cork)
+			corkPowers = append(corkPowers, validatorPower)
+		}
+
+		k.DeleteCork(ctx, val, addr)
```

Comment:
Are there any potential weird side effects of deleting from something you're currently iterating over?

Reply:
Potential yes, but I based this on other cosmos module precommit/validate keeper methods. So I hope its not an issue in many place

Reply:
Not sure if you're referring to the allocation module, but in their winning votes calculation they are iterating over validators and removing commits:

https://github.com/PeggyJV/sommelier/blob/8f0bbc74af38147a25a60974e9ec235541c9abf1/x/allocation/keeper/keeper.go#L391-L419

Reply:
Moreso this:
https://github.com/PeggyJV/sommelier/blob/8f0bbc74af38147a25a60974e9ec235541c9abf1/x/allocation/keeper/keeper.go#L119-L125